rustls/src/msgs/handshake.rs

```diff
@@ -868,7 +868,73 @@ impl TlsListElement for ClientExtension {
     const SIZE_LEN: ListLength = ListLength::U16;
 }
 
+impl TlsListElement for ExtensionType {
+    const SIZE_LEN: ListLength = ListLength::U8;
+}
+
 impl ClientHelloPayload {
+    pub(crate) fn ech_inner_encoding(&self, to_compress: Vec<ExtensionType>) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        self.payload_encode(&mut bytes, Encoding::EchInnerHello { to_compress });
+        bytes
+    }
+
+    pub(crate) fn payload_encode(&self, bytes: &mut Vec<u8>, purpose: Encoding) {
+        self.client_version.encode(bytes);
+        self.random.encode(bytes);
+
+        match purpose {
+            // SessionID is required to be empty in the encoded inner client hello.
+            Encoding::EchInnerHello { .. } => SessionId::empty().encode(bytes),
+            _ => self.session_id.encode(bytes),
+        }
+
+        self.cipher_suites.encode(bytes);
+        self.compression_methods.encode(bytes);
+
+        match purpose {
+            // Compressed extensions must be replaced in the encoded inner client hello.
+            Encoding::EchInnerHello { to_compress } if !to_compress.is_empty() => {
+                // Safety: not empty check in match guard.
```

Comment:
Maybe take this outside of the match arm?

Reply:
Yeah :-/ I had tried a few arrangements but was trying hard to avoid any excessive copying after the clones we definitely require to pull content out of the outer hello for the inner hello.

I'll fiddle to see if there's a nicer way to arrange the match with the first element removal outside the arm.

Reply:
> I'll fiddle to see if there's a nicer way to arrange the match with the first element removal outside the arm.

After some tries I can't come up with an arrangement that won't require more cloning/allocating. I'm going to leave this one as-is unless someone feels strongly.

Reply:
Ohhh! Ok that makes sense. Thank you, I'll fold that in.

Reply:
Done :ballot_box_with_check: